allennlp/modules/matrix_attention/cosine_matrix_attention.py

```diff
@@ -0,0 +1,23 @@
+"""
+A ``Module`` that takes two matrices as input and returns a matrix of attentions.
+"""
+
+import torch
+import torch.nn.functional as F
+from overrides import overrides
+
+from allennlp.common import Params
+from allennlp.modules.matrix_attention.matrix_attention import MatrixAttention
+
+
+@MatrixAttention.register("cosine")
+class CosineMatrixAttention(MatrixAttention):
+
+    @overrides
+    def forward(self, matrix_1: torch.Tensor, matrix_2: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        return F.cosine_similarity(matrix_1, matrix_1)
```

Comment:
@matt-gardner  I am now working on tests and correct implementations for these attention modules. 

I want to compute the cosine similarity of the last dimension in matrix_1 with the cosine similarity of the last dimension in matrix_2 for all entries. 

I tried using:  F.cosine_similarity(matrix_1, matrix_2, dim=-1)

However this only computes the cosine similarity of each entry in matrix_1 with the corresponding entry in matrix_2.

However we want this for all permutations. 
Do you know how to do this without using the expand? 
If there is not a way to do this without using expand then this ticket might not be worth doing.

Reply:
We have an example of this kind of cosine similarity here: https://github.com/allenai/allennlp/blob/9ef013d3a8abee940dbe8e2300f79f8b48cbea42/allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py#L191-L198

And even if some of these implementations don't get any more efficient (e.g., the linear one might not be possible), some of them definitely do, so it's worth it.

Reply:
@matt-gardner @joelgrus 
I think I might have taken the wrong approach originally as I am creating a lot of duplicated code which is bringing no perf gain. 
Instead of having an Implementation of MatrixAttention and Attention called LegacyMatrixAttention and LegacyAttention respectively, maybe we should call these SimilarityFunctionWrappingMatrixAttention and SimilarityFunctionWrappingAttention.
For all the implementations that we can't improve we just leave them use the SimilarityFunctionWrapping* class. 
This way we reduce the amount of duplicated code we are adding.

For all functions we can improve the performance of we will do that in this PR. 

What do you guys think? Also the names of the classes are subject to your approval and if you have better names I'm happy to go with them.